File: rb/features.rb
# DietlyapiIntegration SDK feature factory

require_relative 'feature/base_feature'
require_relative 'feature/ratelimit_feature'
require_relative 'feature/retry_feature'
require_relative 'feature/test_feature'
require_relative 'feature/timeout_feature'


module DietlyapiIntegrationFeatures
  def self.make_feature(name)
    case name
    when "base"
      DietlyapiIntegrationBaseFeature.new
    when "ratelimit"
      DietlyapiIntegrationRatelimitFeature.new
    when "retry"
      DietlyapiIntegrationRetryFeature.new
    when "test"
      DietlyapiIntegrationTestFeature.new
    when "timeout"
      DietlyapiIntegrationTimeoutFeature.new
    else
      DietlyapiIntegrationBaseFeature.new
    end
  end
end

File: py/dietlyapiintegration_sdk/features.py
# DietlyapiIntegration SDK feature factory

from dietlyapiintegration_sdk.feature.base_feature import DietlyapiIntegrationBaseFeature
from dietlyapiintegration_sdk.feature.ratelimit_feature import DietlyapiIntegrationRatelimitFeature
from dietlyapiintegration_sdk.feature.retry_feature import DietlyapiIntegrationRetryFeature
from dietlyapiintegration_sdk.feature.test_feature import DietlyapiIntegrationTestFeature
from dietlyapiintegration_sdk.feature.timeout_feature import DietlyapiIntegrationTimeoutFeature


_FEATURES = {
    "base": lambda: DietlyapiIntegrationBaseFeature(),
    "ratelimit": lambda: DietlyapiIntegrationRatelimitFeature(),
    "retry": lambda: DietlyapiIntegrationRetryFeature(),
    "test": lambda: DietlyapiIntegrationTestFeature(),
    "timeout": lambda: DietlyapiIntegrationTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
